src/GitVersionCore/BuildAgents/GitHubActions.cs

```diff
@@ -24,16 +30,31 @@ public override string GenerateSetVersionMessage(VersionVariables variables)
         public override string[] GenerateSetParameterMessage(string name, string value)
         {
             // https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#environment-files
-            // Example
-            // echo "name=action_state::yellow >> $GITHUB_ENV"
+            // However it's important that GitHub Actions does not parse the log output. The outgoing environment variables must be
+            // written to a temporary file (identified by the $GITHUB_ENV environment variable, which changes for every step in a workflow)
+            // which is then parsed. That file must also be UTF-8 or it will fail.
 
             if (!string.IsNullOrWhiteSpace(value))
             {
-                var key = $"GitVersion_{name}";
+                var gitHubSetEnvFilePath = environment.GetEnvironmentVariable(GitHubSetEnvTempFileEnvironmentVariableName);
+                var assignment = $"GitVersion_{name}={value}";
+
+                if (gitHubSetEnvFilePath != null)
+                {
+                    using (var streamWriter = File.AppendText(gitHubSetEnvFilePath)) // Already uses UTF-8 as required by GitHub
```

Comment:
I still wonder if it would be better to use the base WriteIntegration instead. Now because this is called by the base for each key value pair we end up open/closing file handles N times. With WriteIntegration we could do a select projection of all key value pairs and write all those lines in one go to the file. This would eliminate the N+1 problem

Reply:
Good call @danielmarbach - pushed up a new implementation.